tests/ga/test_agent_update.py

```diff
@@ -0,0 +1,575 @@
+import contextlib
+import json
+import os
+
+from azurelinuxagent.common import conf
+from azurelinuxagent.common.event import WALAEventOperation
+from azurelinuxagent.common.exception import AgentUpgradeExitException, UpdateError
+from azurelinuxagent.common.future import ustr, httpclient
+from azurelinuxagent.common.protocol.restapi import ExtHandlerPackage, VMAgentUpdateStatuses
+
+from azurelinuxagent.common.protocol.util import ProtocolUtil
+from azurelinuxagent.common.version import CURRENT_VERSION, AGENT_NAME
+from azurelinuxagent.ga.agent_update import get_agent_update_handler, GuestAgent, AGENT_MANIFEST_FILE, AGENT_ERROR_FILE, \
+    MAX_FAILURE, GuestAgentError
+from tests.ga.test_update import UpdateTestCase, EMPTY_MANIFEST, WITH_ERROR, NO_ERROR
+from tests.protocol import mockwiredata
+from tests.protocol.HttpRequestPredicates import HttpRequestPredicates
+from tests.protocol.mocks import mock_wire_protocol, MockHttpResponse
+from tests.protocol.mockwiredata import DATA_FILE
+from tests.tools import clear_singleton_instances, load_bin_data, patch
+
+
+class TestAgentUpdate(UpdateTestCase):
+
+    def setUp(self):
+        UpdateTestCase.setUp(self)
+        # Since ProtocolUtil is a singleton per thread, we need to clear it to ensure that the test cases do not
+        # reuse a previous state
+        clear_singleton_instances(ProtocolUtil)
+
+    @contextlib.contextmanager
+    def __get_agent_update_handler(self, test_data=None, autoupdate_frequency=0.001, autoupdate_enabled=True):
+        test_data = DATA_FILE if test_data is None else test_data
```

Comment:
let's default the test_data parameter to DATA_FILE instead of doing it here. It makes the interface more explicit

Reply:
Default to DATA_FILE raising pylint warning
`tests/ga/test_agent_update.py:29:4: W0102: Dangerous default value DATA_FILE (builtins.dict) as argument (dangerous-default-value)`

We can suppress but I decided not to do it unless it's absolutely necessary

Reply:
I see. OK, let's add a doc string describing the default